### PR TITLE
chore: update sigs.k8s.io/karpenter to v1.8.2

### PR DESCRIPTION
**Pull request by tallaxes**
**Description**

Update `sigs.k8s.io/karpenter` to v1.8.2, enable [static capacity](https://karpenter.sh/docs/concepts/nodepools/#specreplicas) support

**How was this change tested?**

* `make presubmit`
* Manual `FOCUS=StaticCapacity make az-upstream-e2etests`. Proper CI integration for testing of this is best done via https://github.com/Azure/karpenter-provider-azure/pull/1363

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

### CODE REVIEW

**Comment by tallaxes** [comment]
Manual run of upstream StaticCapacity passed 10/12, the failures are likely due to metric-server etc. sticky overprovisioning, but have not investigated yet

```
$ FOCUS=StaticCapacity make az-upstream-e2etests
...
Summarizing 2 Failures:
  [FAIL] StaticCapacity Dynamic NodeClaim Interaction [It] should provision dynamic nodeclaims when static nodeclaims cant satisfy pod constraints
  /go/pkg/mod/sigs.k8s.io/karpenter@v1.8.2/test/suites/regression/staticcapacity_test.go:423
  [FAIL] StaticCapacity Edge Cases [It] should handle NodePool deletion gracefully
  /go/pkg/mod/sigs.k8s.io/karpenter@v1.8.2/test/suites/regression/staticcapacity_test.go:507

Ran 12 of 74 Specs in 3486.442 seconds
FAIL! -- 10 Passed | 2 Failed | 0 Pending | 62 Skipped
```

File: charts/karpenter-crd/templates/karpenter.sh_nodepools.yaml [hunk 1]
```diff
@@ -166,8 +169,24 @@ spec:
                       - type: string
                     pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
                     x-kubernetes-int-or-string: true
-                  description: Limits define a set of bounds for provisioning capacity.
+                  description: |-
+                    Limits define a set of bounds for provisioning capacity.
+                    Limits other than limits.nodes is not supported when replicas is set.
```
Comment (matthchr) [inline charts/karpenter-crd/templates/karpenter.sh_nodepools.yaml:174]: minor: I dont think we can fix it here but this isn't correct English, is it?


```suggestion
                    Limits other than limits.nodes are not supported when replicas is set.
```

File: charts/karpenter/templates/deployment.yaml [hunk 2]
```diff
@@ -99,7 +99,7 @@ spec:
                   divisor: "0"
                   resource: limits.memory
             - name: FEATURE_GATES
-              value: "SpotToSpotConsolidation={{ .Values.settings.featureGates.spotToSpotConsolidation }},NodeRepair={{ .Values.settings.featureGates.nodeRepair }},NodeOverlay={{ .Values.settings.featureGates.nodeOverlay }}"
+              value: "SpotToSpotConsolidation={{ .Values.settings.featureGates.spotToSpotConsolidation }},NodeRepair={{ .Values.settings.featureGates.nodeRepair }},NodeOverlay={{ .Values.settings.featureGates.nodeOverlay }},StaticCapacity={{ .Values.settings.featureGates.staticCapacity }}"
```
Comment (matthchr) [inline charts/karpenter/templates/deployment.yaml:102]: Just confirming does this mean that the user can set those values in the CRDs but they'll just be ignored if this feature gate isn't enabled?

File: karpenter-values-template.yaml [hunk 3]
```diff
@@ -57,3 +57,6 @@ serviceAccount:
 podLabels:
   azure.workload.identity/use: "true"
 logLevel: ${LOG_LEVEL}
+settings:
+  featureGates:
+    staticCapacity: true
```
Comment (matthchr) [inline karpenter-values-template.yaml:68]: Why default this to true? We don't default any other alpha features to true, do we? Is this purely for testing sake, since this is our test Helm values file?

Maybe leave a comment to that effect here explaining why we set this to true even though it's alpha.

Comment (matthchr) [reply karpenter-values-template.yaml:68]: and presumably we _won't_ set this to true in the managed NAP experience is that right?

File: charts/karpenter-crd/templates/karpenter.sh_nodepools.yaml (hunk 1 above)
Comment (tallaxes) [reply charts/karpenter-crd/templates/karpenter.sh_nodepools.yaml:174]: No can't fix here and No is not correct English, can submit upstream PR

File: charts/karpenter/templates/deployment.yaml (hunk 2 above)
Comment (tallaxes) [reply charts/karpenter/templates/deployment.yaml:102]: Yes (the static provisioning/deprovisioning controllers are not included)

File: karpenter-values-template.yaml (hunk 3 above)
Comment (tallaxes) [reply karpenter-values-template.yaml:68]: Yes, this is for testing. Will have to figure out the right approach here, especially if using upstream tests; I think some tests expect features to be on (some others maybe flip them dynamically). FWIW, AWS provider flips _all_ features on for development/testing: https://github.com/aws/karpenter-provider-aws/blob/10c4ef30c58b54476f67deff8f62fe81b31c415b/Makefile#L20-L24

(Let's have a separate discussion re state of this in NAP)

**Review by matthchr — DISMISSED** [review]
Approved - feel free to ping me if you push updates and need me to reapprove

**Comment by tallaxes** [comment]
Update on testing: There are three (upstream) tests failing. Two are issues with test code (in one case it loses our NodePool labels/startupTaints causing overprovisioning, another needs a bit of timing adjustment / Eventually to work for Azure provider. The third one however appears to be a real problem: The static provisioner controller does not check if NodePool is being deleted, and can keep provisioning NodeClaims. For some reason this does not surface in AWS testing (I presume). 

All of the above should be fixed by https://github.com/kubernetes-sigs/karpenter/pull/2840; we should hold this until fixes land.

**Comment by tallaxes** [comment]
Temp point to fork for validation. All three previously-failing StaticCapacity tests pass with the tallaxes/karpenter-core fork (branch tallaxes/staticcapacity-fix-v1.8.2). The fork's fixes (DeletionTimestamp check in controller.go, preserving labels/startupTaints from provider NodePool, and Eventually() wrapper in the test) are validated end-to-end against the AKS cluster.

**Comment by tallaxes** [comment]
> All of the above should be fixed by https://github.com/kubernetes-sigs/karpenter/pull/2840; we should hold this until fixes land.

The fix is in 1.10 core, retested on #1585, everything passes
* #1585

**Comment by riyasyash** [comment]
@tallaxes @matthchr Would there be a release of version 1.8.x with these changes or are we expecting a direct jump to 1.10.x from 1.7.2?

**Comment by matthchr** [comment]
A direct jump to 1.10 is the plan.
